### PR TITLE
Improved Near Cache tests

### PR DESCRIPTION
**Pull request by Donnerbart**
The population of the data adapter in Near Cache tests was done during the creation of the `NearCacheTestContext`, because the initial population sends out invalidation messages and we had no method to properly wait for these invalidations to be processed. So we had a race between those invalidations and the following Near Cache population. This lead to a lot of test failures and the solution was to move the data structure population before the creation of the Near Cache instance. This is an unclean design, since the `createContext()` method now also has to take care about the data structure population, which should be a part of the test code.

The existing `NearCacheInvalidationListener` has the problem, that it runs concurrently to the real invalidation listener. So we cannot rely on those `received listener events` to wait (this is again racy and lead to test failures).

This PR introduces a Near Cache statistic for received Near Cache invalidations. There is a public statistic `invalidationCount` which is only increased on removed entries (so after the invalidation listener is processed -> race free). There is also an internal counter `receivedInvalidationCount` which is always increased (which is needed since the initial invalidation doesn't remove any entries). That counter is just accessible via the `NearCacheStatsImpl`, not the interface `NearCacheStats`.

The tests have been changed to use the `populateDataAdapter()` method on their own, which now waits for those initial invalidations. This resolves the race condition without mixing the responsibility of the `createContext()` method with the test code.

This should also speedup the `MapNearCacheBasicTest` a bit, since both members can be created before the data is populated, so there are no more partition migrations when the second member is started.

Some tests had to be changed, since we now don't test received invalidation events, but calls of the remove method. So we might have different numbers in some corner cases (as usual this happened in the lite member and `ReplicatedMap` setup). This might need further investigation, since we might have unwanted duplicated invalidation calls in the code.

The `ClientCacheInvalidationListener` is the only listener, which remains, since the `ClientNearCacheInvalidationTest` relies on asserting the incoming invalidation events, not the number of (removed) entries. The test is much better to read in the old version (I tried to re-write it to the `receivedInvalidationCount` stats, but that was a mess).

Fixes https://github.com/hazelcast/hazelcast/issues/12493
Fixes https://github.com/hazelcast/hazelcast/issues/11456

### CODE REVIEW

**Comment by Donnerbart** [comment]
run-lab-run

**Comment by Donnerbart** [comment]
In a previous version of this PR I disabled the invalidation batching completely, which of course gave the best test performance. But I had the impression that this put more load on the system. So the Near Cache tests were faster, but not the whole PR builder.

Before Sergej's review I switched this to decrease the invalidation batching interval to 1 second (default is 10 seconds). This makes the Near Cache tests a bit slower, but uses less system resources.

Package | Runtime (old) | Runtime (new) | Improvement
-- | -- | -- | --
com.hazelcast.map.impl.nearcache.invalidation | 1 min 15 sec | 1 min 10 sec | 6,67%
com.hazelcast.map.impl.nearcache | 10 min | 7 min 44 sec | 22,67%
com.hazelcast.internal.nearcache.impl.store | 18 ms | 31 ms | -72,22%
com.hazelcast.internal.nearcache.impl.preloader | 1.3 sec | 0.99 sec | 23,85%
com.hazelcast.internal.nearcache.impl.invalidation | 0.29 sec | 1.7 sec | -486,21%
com.hazelcast.internal.nearcache | 9.1 sec | 21 sec | -130,77%
com.hazelcast.client.replicatedmap.nearcache | 29 sec | 32 sec | -10,34%
com.hazelcast.client.map.impl.nearcache.invalidation | 9 sec | 9.6 sec | -6,67%
com.hazelcast.client.map.impl.nearcache | 5 min 37 sec | 6 min 8 sec | -9,20%
com.hazelcast.client.cache.nearcache | 1 min 44 sec | 2 min 37 sec | -50,96%
com.hazelcast.client.cache.impl.nearcache.invalidation | 16 sec | 38 sec | -137,50%
com.hazelcast.client.cache.impl.nearcache | 1 min 5 sec | 1 min 20 sec | -23,08%

The total improvement is 3 seconds (0,27%), so basically we didn't lose any time by waiting for the initial Near Cache invalidations (instead of starting the second instance after the initial population). I think it's still a good trade for the cleaner design.

The configuration is pulled out into the `NearCacheTestUtils`, so we can easily play around with this in the future:
```java
public static Config getBaseConfig() {
  return smallInstanceConfig()  
    .setProperty(CACHE_INVALIDATION_MESSAGE_BATCH_FREQUENCY_SECONDS.getName(), "1")
    .setProperty(MAP_INVALIDATION_MESSAGE_BATCH_FREQUENCY_SECONDS.getName(), "1");
}
```

**Comment by Donnerbart** [comment]
This needs an EE PR, please don't merge until this is done!

**Comment by Holmistr** [comment]
@Donnerbart There are some test failures on PR builder that I haven't seen before, see: https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/15394/
Let me re-run to see how often they appear.

**Comment by Holmistr** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/AbstractNearCacheRecordStore.java [hunk 1]
```diff
@@ -334,23 +334,29 @@ public boolean remove(K key) {
             record = removeRecord(key);
             if (record != null && record.getRecordState() == READ_PERMITTED) {
                 removed = true;
+                nearCacheStats.incrementInvalidations();
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/AbstractNearCacheRecordStore.java]: Should't we call this `incrementEntryRemovalCount`? Other reasons can also be the cause to call this `remove` method like like expirations.

File: hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/AbstractNearCacheRecordStore.java [hunk 2]
```diff
@@ -334,23 +334,29 @@ public boolean remove(K key) {
...
                 nearCacheStats.decrementOwnedEntryCount();
             }
             onRemove(key, record, removed);
             return record != null;
         } catch (Throwable error) {
             onRemoveError(key, record, removed, error);
             throw rethrow(error);
+        } finally {
+            nearCacheStats.incrementReceivedInvalidations();
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/AbstractNearCacheRecordStore.java]: as an addition to the above [comment](https://github.com/hazelcast/hazelcast/pull/12960/files#r184679689) we can call this `incrementInvalidations` (seeing `received` in the name confuses me)

**Comment by Holmistr** [comment]
@ahmetmircik @taburet @Donnerbart So are we good to merge this?

**Comment by Donnerbart** [comment]
@ahmetmircik Are you okay with the PR after my last commit?

File: hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/AbstractNearCacheRecordStore.java [hunk 3]
```diff
@@ -344,13 +344,29 @@ public boolean remove(K key) {
...
+    @Override
+    public boolean invalidate(K key) {
+        try {
+            boolean removed = remove(key);
+            if (removed) {
+                nearCacheStats.incrementInvalidations();
+            }
+            return removed;
+        } finally {
+            nearCacheStats.incrementInvalidationEvents();
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/AbstractNearCacheRecordStore.java]: do all calls to this method are result of invalidation events? there can be also _local_ invalidations triggered after calling methods like imap#remove, imap#put ?

Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/AbstractNearCacheRecordStore.java]: The method doesn't differ between local or remote invalidations, that's true. But I think that's okay, because that's just due to the actual implementation (we could also have local events, it's just a saved overhead to directly call the invalidate method).

But it will not be increased due to evictions or other internal removals, so it's a pure counter for invalidation requests.

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/AbstractNearCacheRecordStore.java]: can we rename it as `incrementInvalidationRequests` instead of `incrementInvalidationEvents` to make it more clear? Other than this i'm ok with this PR.

Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/AbstractNearCacheRecordStore.java]: Done. I've also changed the utility method in `NearCacheTestUtils` accordingly and explained there what an invalidation request includes:
```java
/**
 * Asserts the number of Near Cache invalidation requests (triggered invalidation events and local invalidations).
 *
 * @param context              the given {@link NearCacheTestContext} to retrieve the {@link NearCacheStats} from
 * @param invalidationRequests the given number of Near Cache invalidation requests to wait for
 */
public static void assertNearCacheInvalidationRequests()
```
